47 telegram notification

semjz opened the pull request:
Method send_photo_from_file of TelegramService is not tested cause I don't know how. Also, TelegramSevice might not need 2 methods for sending photo, please check that.

semjz committed:
#47 Create telegram bot service

semjz committed:
Telegram Bot sends message to channel if social_message exists

Added social_message and photo_url keys in ProductController and ProductControllerTest but bot in the Product table.

semjz committed:
#47 Change product tests to not send telegram message

No telegram message will be sent when product tests run because tests shouldn't send any telegram message..

semjz committed:
Update ProductController and ProductControllerTest

TelegramService methods are now called dynamically for test purposes and Bus fake added to ProductControllerTest to test for dispatch of SendProductDetailToTelegram without running the job.

semjz committed:
Update TelegramService and SendProductDetailToTelegram
Make TelegramService method non-static and inject TelegramService dependency in SendProductDetailToTelegram.

semjz committed:
Create tests for TelegramService and SendProductDetailToTelegram

semjz committed:
#47 Update ApiDocumention

semjz committed:
#47 Update TelegramService for testing purposes

Call Constructor inside execute to creat url if it is not created so that when this class is partially mocked it doesn't give error about url not initialized.

semjz committed:
#47 Update QUEUE_CONNECTION value database

semjz committed:
#47 Refactor TelegramService to use laravel HTTP client

semjz committed:
#47 Refactor TelegramServiceTest to fake laravel HTTP client response

semjz committed:
#47 Revert composer.json changes

semjz committed:
#47 Update ApiDocumention

semjz commented:
I will add thumbnail to products table tomorrow please review the code.

geeksesi commented on app/Http/Controllers/API/ProductController.php:
```diff
@@ -22,6 +24,10 @@ public function store(StoreProductRequest $request)
         $data["creator_id"] = auth()->user()->id;
         Product::create($data);
+
+        SendProductDetailToTelegram::dispatchIf(isset($data["social_message"])
```
Fix DRY 
https://www.youtube.com/watch?v=8DBMRByJUmU

semjz replied:
Added SendProductDetailToTelegramAction and used dispatch inside of it, Action will be called in the controller.

geeksesi commented on app/Http/Controllers/API/ProductController.php:
```diff
@@ -32,7 +38,12 @@ public function show(Product $product)
+        $data = $request->validated();
+        $product->update($data);
+
+        SendProductDetailToTelegram::dispatchIf(isset($data["social_message"])
```
fix DRY here

semjz replied:
Added SendProductDetailToTelegramAction and used dispatch inside of it, Action will be called in the controller.

geeksesi commented on app/Jobs/SendProductDetailToTelegram.php:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @return void
+     */
+    public function __construct($telegram_service ,$photo_url, $title, $description)
```
for telegram service use laravel dependency injection ( service container )

semjz replied:
Added TelegeramService to parameters of handle function .It adds TelegeramService  as a dependency injection.

geeksesi commented on app/Services/TelegramService.php:
```diff
@@ -0,0 +1,74 @@
+     */
+    public function send_photo_from_file(string $photo_path, string $chat_id,string $title, string $description):array|bool
+    {
+        $photo = fopen($photo_path, 'r');
```
use file_get_content instead, it's not safe.

geeksesi replied:
or even use laravel storage.
it's not safe to open file like this.

semjz replied:
I can't resolve this because it seems "multipart/form-data" wants the resource pointer of the file I guess? so it gives a "Bad Request" error when using file_get_content. So, we need to talk about this.

geeksesi replied:
find solution, laravel has solution for this.

geeksesi commented on tests/Feature/Http/Controllers/API/ProductControllerTest.php:
```diff
@@ -43,21 +45,30 @@ public function testDeleteProduct()
 
     public function testCreateProduct()
     {
+        Bus::fake();
```
Why Bus ?! you have to use Queue:fake

semjz replied:
I fixed this by using Queue:fake which only tests if SendProductDetailToTelegram job is pushed to the queue.

semjz committed:
#47 Add thumbnail to product table, model and requests

semjz committed:
#47 Create action to clear up ProductController

semjz committed:
#47 Use queue::fake instead of Bus:fake
To check if the job is pushed in the queue.

semjz committed:
#47 Create SendProductDetailToTelegramAction

semjz committed:
#47 Add dependency injection in handler method

semjz committed:
#47 Test with the added dependency injection

semjz committed:
#47 Add thumbnail in Product requests

semjz committed:
#47 Just some variable name changes

semjz commented:
Also, Do we really need SendProductDetailToTelegramTest.php? and if so please check this file I don't think the code is correct.

semjz committed:
#47 Update ApiDocumention

semjz committed:
#47 Fake http response for the actual used url in TelegramService

geeksesi commented on app/Http/Controllers/API/ProductController.php:
```diff
@@ -30,9 +34,11 @@ public function show(Product $product)
     }
 
-    public function update(UpdateProductRequest $request, Product $product)
+    public function update(UpdateProductRequest $request, Product $product, SendProductDetailToTelegramAction $action)
```
also this should be in constructor

geeksesi commented on app/Http/Controllers/API/ProductController.php:
```diff
@@ -17,11 +20,12 @@ public function index(Request $request)
     }
 
-    public function store(StoreProductRequest $request)
+    public function store(StoreProductRequest $request, SendProductDetailToTelegramAction $action)
```
this should be in Constroctuor instead of here.

geeksesi commented on app/Http/Requests/API/ProductController/StoreProductRequest.php:
```diff
@@ -32,6 +32,8 @@ public function rules()
             'price' => ['nullable'],
             'type' => ['required', 'string', new Enum(ProductTypeEnum::class)],
+            'social_message' => ['nullable', 'string'],
+            'thumbnail' => ['nullable', 'string']
```
thumbnail should be file. not string.!!

geeksesi commented on database/factories/ProductFactory.php:
```diff
@@ -27,7 +27,8 @@ public function definition()
             'price' => 2000,
-            'type' => ProductTypeEnum::DIGITAL
+            'type' => ProductTypeEnum::DIGITAL,
+            'thumbnail' => fake()->url
```
use storage fake instead of fake url.

geeksesi commented on database/migrations/2023_02_06_074026_add_thumbnail_to_products_table.php:
```diff
@@ -0,0 +1,32 @@
+    public function up()
+    {
+        Schema::table('products', function (Blueprint $table) {
+            $table->string('thumbnail');
```
you made it nullable in form request validation, but in database it's not null.!!

geeksesi commented on tests/Feature/Jobs/SendProductDetailToTelegramTest.php:
```diff
@@ -0,0 +1,30 @@
+            ->once()
+            ->with($photo_url, $chat_id , $title, $description);
+
+        $SendProductDetailToTelegram = new SendProductDetailToTelegram($photo_url, $title, $description);
```
variable names are camelCase. it's CapitalCase. fix it please :)

semjz committed:
#47 Clean up

semjz committed:
#47 Fake Storage and upload file for testing purposes

semjz committed:
#47 Change TelegramService to Upload the photo Using Storage

Discarded usage of fopen()

semjz committed:
#47 Thumbnail has to be a file

semjz committed:
#47 Use Product factory in TelegramServiceTest and Use correct method

semjz committed:
#47 Update ApiDocumention

semjz committed:
#47 merge master into telegram-notification

semjz committed:
#47 Add action in store method to pass the creatProduct test

semjz committed:
#47 Update ApiDocumention

semjz committed:
#47 Upload Photo and save full path of it in products table

semjz committed:
#47 Tidy up ProductFactory and TelegramService

Storage::fake() doesn't do anything,so it is removed.
TelegramService sends photo via files so we only need multipart/form-data.

semjz committed:
#47 Change test so is actually mocks TelegramService
Multipart/form-data requires the file to be a resource object otherwise it will send a 400 code ("Bad Request"), so if in our code execute function receives a resource object we can assume that the response will be a 200 code otherwise it is 400.

semjz committed:
#47 Change variable name

semjz committed:
#47 Store the relative path of photo in products table in regard to Storage file

semjz committed:
#47 Use fake storage for testing TelegramService
